### PR TITLE
C bindings: pass context to set_properties and warp operation callbacks

### PR DESCRIPTION
**Pull request by harmanpa**
Similar to #427 , this PR creates and tests functions in c bindings for passing data to the callback function. We use Manifold from Java and this allows us to pass callbacks from Java that carry references to the associated Java objects, however it is also useful for passing parameters, as done in the tests in manifoldc_test.cpp.

New functions:
* manifold_set_properties_context
* manifold_warp_context
* manifold_cross_section_warp_context

New test:
* CBIND properties

Changed tests:
* CBIND warp_translation
* CBIND level_set

### CODE REVIEW

**Comment by pca006132** [comment]
Do we have more functions that accept a callback without a ctx parameter? I think this will be the nice place to fix everything (if there is any). I think we don't have to introduce a `context` variant for these APIs, to avoid the API from being too complicated.

The problem here is *will it cause existing problems to break badly*? I think it should either cause compilation failure or not affecting existing program. I think adding parameters at the end of the list of arguments is fine. For x86-64, we are looking at `cdecl` calling convention, which is fine with additional parameters (it will just ignore the additional one at the end) or fewer parameters (it will just be junk). For AArch64, it should also be fine as the parameters are passed in register.

TL;DR: It is fine to just add parameters at the end. Standard C API is fine with this.

**Comment by harmanpa** [comment]
I couldn't find other methods in the C bindings with a callback.

Your suggestion is to have the context parameter always but it be optional?

File: src/manifold/src/sort.cpp [hunk 1]
```diff
@@ -464,6 +464,7 @@ void Manifold::Impl::GatherFaces(const Impl& old, const Vec<int>& faceNew2Old) {
     gather(policy, faceNew2Old.begin(), faceNew2Old.end(),
            old.meshRelation_.triProperties.begin(),
            meshRelation_.triProperties.begin());
+    meshRelation_.numProp = old.meshRelation_.numProp;
```
Comment (elalish) [inline src/manifold/src/sort.cpp:467]: It looks like you may have started with your other PR's branch. Should probably rebase onto master to avoid merge conflicts - and to make it easier to review.

File: bindings/c/manifoldc.cpp [hunk 2]
```diff
@@ -275,6 +275,21 @@ ManifoldManifold *manifold_warp(void *mem, ManifoldManifold *m,
   return to_c(new (mem) Manifold(warped));
 }
 
+ManifoldManifold *manifold_warp_context(void *mem, ManifoldManifold *m,
```
Comment (elalish) [inline bindings/c/manifoldc.cpp:278]: I agree that if adding an optional ctx parameter to a single function works, that does seem cleaner. It would mean a small breaking change for the existing level set function and such, but since I'm guessing you're the only user of this for now, it's probably fine.

File: test/manifoldc_test.cpp [hunk 3]
```diff
@@ -23,19 +23,35 @@ TEST(CBIND, warp_translation) {
...
+  ManifoldManifold *warpedcontext =
+      manifold_warp_context(malloc(sz), sphere, warpcontext, context);
+  ManifoldManifold *diffcontext =
+      manifold_difference(malloc(sz), trans, warped);
 
   ManifoldProperties props = manifold_get_properties(diff);
+  ManifoldProperties propscontext = manifold_get_properties(diffcontext);
 
   EXPECT_NEAR(props.volume, 0, 0.0001);
+  EXPECT_NEAR(propscontext.volume, 0, 0.0001);
```
Comment (elalish) [inline test/manifoldc_test.cpp:46]: If what you're testing is translation, perhaps we should add a bounding box check to verify it moved.

File: test/manifoldc_test.cpp [hunk 4]
```diff
@@ -69,10 +106,41 @@ TEST(CBIND, level_set) {
 #endif
 
   EXPECT_EQ(manifold_status(sdf_man), MANIFOLD_NO_ERROR);
+  EXPECT_EQ(manifold_status(sdf_man_context), MANIFOLD_NO_ERROR);
```
Comment (elalish) [inline test/manifoldc_test.cpp:109]: Can we add some volume and surface area checks to tighten these tests?

**Review by elalish — CHANGES_REQUESTED** [review]
Thanks! Good idea.

**Comment by pca006132** [comment]
> Your suggestion is to have the context parameter always but it be optional?

Kind of, users should put a null if they don't want to use it (and are aware of this API update).

My explanation above is just me thinking about if this will break applications if they assume the old API. Because linker does not check the number of arguments, they may not notice the API change. And my conclusion is it is probably fine.

**Comment by harmanpa** [comment]
Sorry for all the commits, I'm more used to C than C++ so not knowing what libraries to use.

I've made the "context" variants of functions into the defaults. All builds and tests but don't know possible external impacts.

File: test/manifoldc_test.cpp [hunk 5]
```diff
@@ -19,29 +19,76 @@ TEST(CBIND, sphere) {
...
+  context[0] = 15.0f;
+  ManifoldVec3 (*warpcontext)(
+      float, float, float, void *) = [](float x, float y, float z, void *ctx) {
+    ManifoldVec3 v = {x + ((float *)ctx)[0], y, z};
+    return v;
+  };
   ManifoldManifold *sphere = manifold_sphere(malloc(sz), 1.0f, 100);
   ManifoldManifold *trans = manifold_translate(malloc(sz), sphere, 15., 0., 0.);
-  ManifoldManifold *warped = manifold_warp(malloc(sz), sphere, warp);
+  ManifoldManifold *warped = manifold_warp(malloc(sz), sphere, warp, NULL);
```
Comment (elalish) [inline test/manifoldc_test.cpp:36]: Can we test @pca006132's idea by simply omitting `NULL` here (and in the other tests)? If that works, then this is a pretty clean API. The only breaking change is removing `manifold_level_set_context`, which I'm comfortable with, especially since you're the main user of it.

File: test/manifoldc_test.cpp [hunk 6]
```diff
@@ -69,10 +137,70 @@ TEST(CBIND, level_set) {
...
+        5.0f;
+  };
+  float *context = (float *)malloc(1 * sizeof(float));
+  context[0] = 5.0f;
+  void (*propscontext)(float *, ManifoldVec3, const float *,
+                       void *) = [](float *new_prop, ManifoldVec3 position,
+                                    const float *old_prop, void *ctx) {
+    new_prop[0] =
+        glm::sqrt(glm::sqrt(position.x * position.x + position.y * position.y) +
+                  position.z * position.z) *
```
Comment (elalish) [inline test/manifoldc_test.cpp:189]: How about `glm::length(position)`?

File: test/manifoldc_test.cpp [hunk 7]
```diff
@@ -69,10 +137,70 @@ TEST(CBIND, level_set) {
...
+
+  // Analytic calculations for volume and surface area
+  float a = context[0] * context[1];
+  float b = context[0] * context[2];
+  float c = context[0] * context[3];
+  float s = 4.0f * glm::pi<float>() *
+            std::pow(((std::pow(a * b, 1.6f) + std::pow(a * c, 1.6f) +
+                       std::pow(b * c, 1.6f)) /
+                      3.0f),
+                     1.0f / 1.6f);
```
Comment (elalish) [inline test/manifoldc_test.cpp:150]: Thanks for the geometry lesson! 😄

**Review by elalish — COMMENTED** [review]
I assume you've tested this is all working with your Java bindings?

File: test/manifoldc_test.cpp (hunk 5 above)
Comment (harmanpa) [reply test/manifoldc_test.cpp:36]: Unfortunately I can't get it to compile without the NULL in the tests.

I know there's tricks we can play with varargs or preprocessor but not sure how you would feel about that.

Comment (elalish) [reply test/manifoldc_test.cpp:36]: No, that's fine, let's leave it. I assume C doesn't have nice things like default arguments? @pca006132 Did I misunderstand what you were saying?

Comment (pca006132) [reply test/manifoldc_test.cpp:36]: Yeah I mean something else. Will give an example later today.

**Comment by pca006132** [comment]
This is what I mean:

```shell
$ cat foo.c  
#include <stdio.h>

int foo(int a, int (*cb)(int, int), int b) {
  printf("%d %d %d\n", a, b, cb(a, b));
  return 0;
}

$ cat bar.c
#include <stdio.h>

int foo(int a, int (*cb)(int));

int cb(int a) {
  printf("a = %d\n", a);
  return a * 10;
}

int main() {
  printf("%d\n", foo(1, cb));
  return 0;
}

$ gcc foo.c bar.c
$ ./a.out
a = 1
1 -1245249192 10
0
```

i.e. when the users are not using the header file from C but declares the function in their language, it should not affect the original functionality.

File: test/manifoldc_test.cpp (hunk 5 above)
Comment (elalish) [reply test/manifoldc_test.cpp:36]: Okay, i guess we can skip this one. Is this ready to merge then?

Comment (harmanpa) [reply test/manifoldc_test.cpp:36]: I'm happy if you're happy
